warehouse/redshift.go

```diff
@@ -330,12 +365,13 @@ func (rs *Redshift) RemoveOrphanedRecords(lastSync pq.NullTime) error {
 	return nil
 }
 
-// DoesTableExist checks if a table with a given name exists in the public schema
+// DoesTableExist checks if a table with a given name exists
 func (rs *Redshift) DoesTableExist(name string) bool {
 	log.Printf("Checking if table %s exists", name)
 
 	var exists int
-	err := rs.conn.QueryRow("SELECT count(*) FROM pg_tables WHERE schemaname = 'public' AND tablename = $1;", name).Scan(&exists)
+	query := fmt.Sprintf("SELECT count(*) FROM pg_tables WHERE schemaname = %s AND tablename = '%s';", rs.getSchemaParameter(), name)
```

Comment:
nit: Here and in `getTableColumns` below, I think the `$1` query parameter for the table name was preferable (rather than sprintf-ing literals into the SQL string) to prevent SQL injection. Given that the schema may sometimes be a function call, I guess we can't use a parameter for that.

The way we use `hauser` makes it very unlikely of course that there will be any avenues for actual SQL injection, but when it doesn't make the code awkward, I prefer to keep protecting against it. You never know how this code might evolve or who might copy/paste sections, so it's best that it be bullet-proof.

Reply:
I totally agree. String concatenation is used to build SQL queries throughout redshift.go (insert, create, and update statements). The two statements that used parametrized SQL were the `DoesTableExist` and `getTableColumns` functions. A deeper refactor is in order to parameterize all queries. Until then, I'll add $1 back to the two statements that I changed.

Reply:
In the process of updating this, I found an issue with the query in `getTableColumns` that prevented columns for existing tables using arbitrary schemas from being fetched. The original query used the `PG_TABLE_DEF` table, which will only return tables for schemas that are in the search_path (see https://docs.aws.amazon.com/redshift/latest/dg/r_PG_TABLE_DEF.html). This wasn't apparent when all tables were created using the "public" schema (which is in the search_path by default).

I fixed this by querying system tables under the `information_schema` schema, which is actually a SQL standard: https://www.postgresql.org/docs/9.1/information-schema.html

Reply:
Nice, thanks!